Complexity_classes_def/P.thy

```diff
@@ -331,4 +343,140 @@ definition NP_hard :: "lang set" where
 
 definition NP_complete :: "lang set" where
 "NP_complete \<equiv> NP_hard \<inter> NP"
-  
+
+locale encode_decode_sat =
+  fixes encode_sat :: "nat three_sat \<Rightarrow> nat"
+  fixes decode_sat :: "nat\<Rightarrow> nat three_sat"
+  assumes decode_encode_inv : "decode_sat (encode_sat F) = F"
+begin
+definition IMP_SAT :: "nat set" where 
+"IMP_SAT == encode_sat ` {n. sat n}"
+
+lemma main_lemma : 
+  fixes c p
+  assumes "poly p"
+  assumes "\<forall>s. \<exists>t. Ex (big_step_t (c, s) t) \<and> t \<le> p (length (s ''input''))"
+  assumes " \<forall>x z. \<exists>r. \<forall>s. s ''input'' = x \<and> s ''certificate'' = z \<longrightarrow>
+                         (\<exists>t s'. (c, s) \<Rightarrow>\<^bsup> t \<^esup> s' \<and> s' ''input'' = r)"
+  shows "\<exists>imp_to_sat t_red s_red.
+         poly t_red 
+       \<and> poly s_red
+       \<and> (\<forall>s t s'. (imp_to_sat,s) \<Rightarrow>\<^bsup> t \<^esup> s' \<longrightarrow> s ''certificate'' = s' ''certificate'')
+       \<and> (\<forall>x. \<exists>f.    length f \<le> s_red ( length x )
+                   \<and> (\<forall>s. s ''input'' = x \<longrightarrow> (\<exists>t s'. (imp_to_sat,s) \<Rightarrow>\<^bsup> t \<^esup> s' \<and> 
+                                          s' ''input'' = f \<and> t \<le> t_red(length x)))
+                   \<and> ( f \<in> IMP_SAT \<longleftrightarrow>
+                                        (\<exists>z. \<forall>s t s'. s ''input'' = x 
+                                                      \<and> s ''certificate'' = z 
+                                                      \<and>(v, s) \<Rightarrow>\<^bsup> t \<^esup> s' \<longrightarrow> s' ''input'' >0)
+                                        )
+                      )"
+  sorry
```

Comment:
@maxhaslbeck thoughts about this version of this main-lemma ?

Reply:
obviously you have a typo: `v` and `c` in `main_lemma` should be the same.
The proof runs through nonetheless. Good work!

Okay, I'll try to phrase your `main_lemma` in natural language:

Let `c` be a `IMP-` program and `p :: nat => nat` a growth function with the following properties:
- `p` is polynomially bounded
- `c`terminates on every input state `s`, with running time bounded by `p N`, where N is the size of the input.
- the computation of `c` only depends on the initial values of registers ''input'' and ''certificate''
Then, there is an `IMP-` program `imp_to_sat` that computes from any input `x` a Propositional Formula `f(x)` that is SAT iff **there exists a certificate** `z` such that c[x,z] returns true.
Formally,
- two growth functions ensure time (`t_red`) and space (`s_red`) consuption of the reduction
- `imp_to_sat` leaves the register ''certificate'' untouched
- the size of the formula `f` is bounded in the size of the input
- `imp_to_sat` computes `f` in time bounded by `t_red`
- `f`is SAT iff **there exists a certificate** `z` such that c[x,z] returns true.

Questions:
- the running time of `c` is independent of the size of the certificate. Whats the consequence of that? artificially padding the certificate is made impossible by that?
- thinking from the other end I suspect that the `there exists a z` should not be in the main lemma. the role of `imp_to_sat` is to put the deterministic computation into a propositional formula. The work of finding the satisfying assignment/the certificate should be included in the reduction, does it? @notiho, @mabdula what do you think about it?

Minor comment:
- the name `length` may be misleading, and confused with the length of a list.

Reply:
Having thought about it a bit more, I think the `there exists a certificate` is fine. It means, that in the final propositional formula, there must be popositional variables for the input 'certificate', which the SAT solver then has to find.

Reply:
First Thanks .

Your natural language interpretation of the main-lemma is what it is was expected to be. This should be a good sign ! 
I'd add only one thing item to the results to be shown: the result of `imp_to_sat` does only depend of the `input` register. This is kind of mentioned in the same statement that says that  the computation of `f(x)` is bounded by `t_red`  (just like the code c depending only on `input` and `certificate`) 
 
Answering the questions: 
- I am not really sure I understand what do you mean by padding the certificate. But I'll explain the motivation behind excluding the size of the certificate. the literature (that I know of) uses only the input for defining polynomial verifiers and thus NP. See for instance Definition 6.7 of the [slides of Einführung in die Theoretische Informatik](https://www21.in.tum.de/teaching/theo/SS20/folien-handout.pdf)  .
And it kind of makes sense. Had we included the certificate in the time bound function, then a problem like [PCP](https://en.wikipedia.org/wiki/Post_correspondence_problem) would be in NP. since a certificate could be a solution of the PCP instance. The instance with the certificate is verifiable in polynomial time wrt to the length of the certificate. This is of course wrong proof that PCP is in NP, because we know for instance that PCP is not even decidable. The flaw in the proof is that certificate could get indefinitely long. In the same reasoning, also the halting problem would be NP. As  the existence of a sequence of small step configurations / TM configurations. In order to do that we should either require a bound on the size of certificate of belonging to the set wrt to the size of input, or exclude the certificate size as I did here.

- For the second question, that is exactly what I was going to say (what you later said). The reduction should by definition only take the input x and gives us a formula. So I guess there's no avoiding  `there exists a certificate`. So the formula will kind have to include the constraints you should have on a certificate. So the existence of the assignment means the existence of a certificate. (Although I wish we could just include finding the certificate in the reduction. That would be a literally a million dollar achievement  )

PS: what do you recommend changing the name of `length` to ?
PS: Yes that was a typo and I fixed it now. I freaked out when I realised that but thank god it had miraculously no effect on the correctness of the proof.

Reply:
Back on the issue of the first question and the definition of NP in general. 

Thank you for your question about that issue because I think it is an important search topic for the validity of our definitions and models. 

It looks that for example that "The nature of computation" does suggest a definition of NP that uses the other alternative as what I did. It includes the certificate in the running time bound, but puts a bound on the certificate. I pondered a little bit about whether the 2 definitions are equivalent. Going from the certificate length bounded definition to the one that is not is trivial for every model of computation. The other direction works on a TM since having a time limit always mean a space limit. So a very long certificate is to be implicitly truncated. Our IMP- model doesn't guarantee this condition in general. Yet we know something about our well formed verifiers. The result only depends of the input and the certificate. So for a fixed input , there exists some "maximal certificate" which is a number from which every certificate behaves like infinity (Since the input is fixed and everything else is a constant). this maximal certificate is polynomially bound for sure. I do believe nonetheless that a proof of that would not be really easy. Direct induction alone wouldn't be much helpful since when a program's result doesn't depend of some variable. This doesn't mean it is the case for every part of it. 

So I say for the moment, we do pick the definition that will help the most in proving the main-lemma. If the reduction will need as info that a certificate if exists has to be bounded or that time bound depends from the certificate length. Then one could edit the definition of NP to explicitly mention that and do the changes that need to be done. If not, I say we stick with the actual one. It seems the one keeping us from much trouble. 

A really better scenario would be being able to prove that our definition of NP is enough to put an upper bound on the certificate. But I am not sure we do have the infrastructure for it.

Reply:
@maxhaslbeck  @notiho  @mabdula  . Did update the main-lemma so that the reduction says (not directly ofc but through the resulting SAT-formula) for a well known and fixed polynomial upper bound , whether a certificate that is bounded with aforementioned bound works with the given x. 

Merry Christmas and I am looking forward to your feedback when available again !

Reply:
PS: main-lemma is in the last commit line 403 of P.thy (the only file I'm working on , on the moment)